### PR TITLE
Fix method call semantics

### LINKED ISSUE #19: Virtual Call Resolution does not work correctly for private methods
Taken from Bitbucket#179:

ProjectLike's virtualCall resolution does not work in the following code evolution scenario. The JVM runs this code (please see the attached files. B has a main).

Class Structure
```
class A {
   protected void method(){\* do  something*\}
}

class B extends A {
   private void method(){/* do something*/}

  public static void main(String ... args) {
    var B = new B();
    b.method(); // this is an INVOKEVIRTUAL
  }
}
```
**Problem**: b.method is resolved to A.method when an invokevirtual is used.

sbt console code leading to this error:

```
val file = new java.io.File(dir)
val project = org.opalj.br.analyses.Project(file)
val methods = project.allMethodsWithBody
val main = methods.filter(_.name =="main")
val body = main.body.get
val invoke = body.collectFirst{
   case x @ INVOKEVIRTUAL(org.opalj.br.ObjectType("B"),"m", _) => x
}.get
val results = project.virtualCall("", invoke)
```
>OUTPUT: results: scala.collection.Set[org.opalj.br.Method] = Set(protected void m())

The respective classfiles are attached. "StrangeCallCreator.scala" is the DSL script that engineered these classes.

[files.zip](https://github.com/stg-tud/opal/files/4029707/files.zip)

### LINKED ISSUE #22: Assertion in DeclaredMethodsKey
```
Assertion Error in com.botbrew.basil_24-AppTrackerEncryptor.jar Exception in thread "main" org.opalj.concurrent.ConcurrentExceptions: concurrent exceptions occurred Suppressed: java.lang.AssertionError: assertion failed: creation of declared methods failed: DefinedMethod(declaringClassType=com.viewpagerindicator.IcsLinearLayout,definedMethod=com.viewpagerindicator.IcsLinearLayout{ private boolean hasDividerBeforeChildAt(int) }) vs.(new) DefinedMethod(declaringClassType=com.viewpagerindicator.IcsLinearLayout,definedMethod=android.widget.LinearLayout{ protected boolean hasDividerBeforeChildAt(int) })} at scala.Predef$.assert(Predef.scala:223) at org.opalj.br.analyses.DeclaredMethodsKey$.insertDeclaredMethod$1(DeclaredMethodsKey.scala:109) at org.opalj.br.analyses.DeclaredMethodsKey$.$anonfun$compute$14(DeclaredMethodsKey.scala:200) at org.opalj.br.analyses.DeclaredMethodsKey$.$anonfun$compute$14$adapted(DeclaredMethodsKey.scala:197) at org.opalj.collection.immutable.ConstArray.foreach(ConstArray.scala:40) at org.opalj.br.analyses.DeclaredMethodsKey$.$anonfun$compute$5(DeclaredMethodsKey.scala:197) at org.opalj.br.analyses.DeclaredMethodsKey$.$anonfun$compute$5$adapted(DeclaredMethodsKey.scala:114) at org.opalj.concurrent.package$.$anonfun$parForeachArrayElement$2(package.scala:223) at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23) at scala.concurrent.Future$.$anonfun$apply$1(Future.scala:658) at scala.util.Success.$anonfun$map$1(Try.scala:255) at scala.util.Success.map(Try.scala:213) at scala.concurrent.Future.$anonfun$map$1(Future.scala:292) at scala.concurrent.impl.Promise.liftedTree1$1(Promise.scala:33) at scala.concurrent.impl.Promise.$anonfun$transform$1(Promise.scala:33) at scala.concurrent.impl.CallbackRunnable.run(Promise.scala:64) at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) at java.lang.Thread.run(Thread.java:748)
```

See Bitbucket for files to reproduce.

### PR DESCRIPTION
**Pull request by errt**
All invocation instructions can actually invoke private methods. This PR changes the semantics of Project.instanceMethods to include these private methods and adapts the call resolution functions of ProjectLike to work accordingly.
Additionally, accessibility checks are now performed in ProjectLike, no client ever implemented them anyway although ProjectLike cautioned to do so.

This fixes #19. It also fixes #22.

### CODE REVIEW

File: OPAL/br/src/main/scala/org/opalj/br/Method.scala [hunk 1]
```diff
@@ -465,6 +465,19 @@ final class Method private[br] (
...
     override def asMethod: this.type = this
+
+    def isAccessibleBy(objectType: ObjectType)(implicit classHierarchy: ClassHierarchy): Boolean = {
+        visibilityModifier match {
+            case Some(ACC_PUBLIC) ⇒ true
+            case Some(ACC_PROTECTED) ⇒
+                declaringClassFile.thisType.packageName == objectType.packageName ||
+                    objectType.isASubtypeOf(declaringClassFile.thisType).isNotNo
+            case Some(ACC_PRIVATE) ⇒
+                // TODO Java 11+ allows access to private methods from nestmates
```
Comment (TorunR) [inline OPAL/br/src/main/scala/org/opalj/br/Method.scala:476]: Did you miss this or is that a TODO for later ?

Comment (errt) [reply OPAL/br/src/main/scala/org/opalj/br/Method.scala:476]: Left to fix in a separate pull request, namely #62

File: DEVELOPING_OPAL/validate/src/test/scala/org/opalj/fpcf/ThrownExceptionsAnalysisTests.scala [hunk 2]
```diff
@@ -29,11 +29,8 @@ class ThrownExceptionsAnalysisTests extends PropertiesTest {
         List("org/opalj/fpcf/fixtures/thrown_exceptions")
     }
 
-    describe("L1ThrownExceptionsAnalysis and EagerVirtualMethodAllocationFreenessAnalysis are executed") {
-        val as = executeAnalyses(Set(
-            EagerVirtualMethodAllocationFreenessAnalysis,
-            EagerL1ThrownExceptionsAnalysis
-        ))
+    describe("L1ThrownExceptionsAnalysis are executed") {
+        val as = executeAnalyses(Set(EagerL1ThrownExceptionsAnalysis))
```
Comment (TorunR) [inline DEVELOPING_OPAL/validate/src/test/scala/org/opalj/fpcf/ThrownExceptionsAnalysisTests.scala]: Why do you no longer need that: Or rather the other way around Why did you need that ever?

File: OPAL/br/src/main/scala/org/opalj/br/analyses/Project.scala [hunk 3]
```diff
@@ -1458,6 +1461,21 @@ object Project {
                                 staticallyOverriddenInstanceMethods ::=
                                     ((objectType, declaredMethodName, declaredMethodDescriptor))
                             }
+                        } else if (!declaredMethod.isInitializer) {
+                            // Private methods can be invoked by invokevirtual instructions (and
+                            // invokeinterface for Java 11+). If a call is resoved to a private
```
Comment (TorunR) [inline OPAL/br/src/main/scala/org/opalj/br/analyses/Project.scala]: resoved -> resolved

File: DEVELOPING_OPAL/validate/src/test/scala/org/opalj/fpcf/ThrownExceptionsAnalysisTests.scala (hunk 2 above)
Comment (errt) [reply DEVELOPING_OPAL/validate/src/test/scala/org/opalj/fpcf/ThrownExceptionsAnalysisTests.scala]: I asked that myself. But now I found the reason for this: This seems to be a botched renaming. I'll change this to the analysis that was supposed to be there.

**Review by TorunR — COMMENTED** [review]
I think the typo and that strange Domain are open

**Review by TorunR — APPROVED** [review]

**Pull request merged**